cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java

```diff
@@ -271,6 +280,56 @@ protected boolean propagateServiceError() {
     return true;
   }
 
+  /**
+   * Creates a {@link ProgramRunner} for the running the program in this {@link TwillRunnable}.
+   */
+  protected ProgramRunner createProgramRunner(Injector injector) {
+    Type type = TypeToken.of(getClass()).getSupertype(AbstractProgramTwillRunnable.class).getType();
+    // Must be ParameterizedType
+    Preconditions.checkState(type instanceof ParameterizedType, "Invalid ProgramTwillRunnable class %s.", getClass());
+
+    Type programRunnerType = ((ParameterizedType) type).getActualTypeArguments()[0];
+    // the ProgramRunnerType must be a Class
+    Preconditions.checkState(programRunnerType instanceof Class,
+                             "ProgramRunner type is not a class: %s", programRunnerType);
+
+    @SuppressWarnings("unchecked")
+    Class<ProgramRunner> programRunnerClass = (Class<ProgramRunner>) programRunnerType;
+    return injector.getInstance(programRunnerClass);
+  }
+
+  /**
+   * Creates a {@link Program} for the given {@link ProgramRunner} from the given program jar {@link File}.
+   */
+  private Program createProgram(ProgramRunner programRunner, File programJarFile) throws IOException {
+    File unpackedDir = Files.createTempDir();
+
+    FilterClassLoader.Filter filter;
+    if (programRunner instanceof ProgramClassLoaderFilterProvider) {
+      filter = ((ProgramClassLoaderFilterProvider) programRunner).getFilter();
```

Comment:
This will allow the provider to ignore `defaultFilter()` if they want to. Is that expected usage?